File: app/src/main/java/com/giosis/util/qdrive/list/SigningView.java
package com.giosis.util.qdrive.list;


import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Path;
import android.util.AttributeSet;
import android.util.Log;
import android.view.MotionEvent;
import android.view.View;

import com.giosis.util.qdrive.util.DisplayUtil;

public class SigningView extends View {

    public SigningView(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
        init(context);
    }

    private Bitmap mBitmap;
    private Canvas mCanvas;
    private Path mPath;
    private Paint mBitmapPaint;
    private Paint mPaint;

    private boolean bIsClear = false;
    private boolean bIsDrawStart = false;
    private boolean bIsTouche = false;

    private Context mContext;

    public SigningView(Context context) {
        super(context);
        init(context);
    }

    public void setIsTouche(boolean status) {
        this.bIsTouche = status;
    }

    public boolean getIsTouche() {
        return this.bIsTouche;
    }


    public SigningView(Context context, AttributeSet attrs) {
        super(context, attrs);
        init(context);
    }

    private void init(Context context) {
        Log.e("krm0219", "SingingView init");
        mPaint = new Paint();
        mPaint.setAntiAlias(true);
        mPaint.setDither(true);
        mPaint.setColor(0xFF000000);
        mPaint.setStyle(Paint.Style.STROKE);
        mPaint.setStrokeJoin(Paint.Join.ROUND);
        mPaint.setStrokeCap(Paint.Cap.ROUND);
        mPaint.setStrokeWidth(8);

        int width = DisplayUtil.DPFromPixel(context, 480);
        int height = DisplayUtil.DPFromPixel(context, 480);

        mBitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888);
        mCanvas = new Canvas(mBitmap);
        mPath = new Path();
        mBitmapPaint = new Paint(Paint.DITHER_FLAG);

        mContext = context;
    }

    @Override
    protected void onSizeChanged(int w, int h, int oldw, int oldh) {
        super.onSizeChanged(w, h, oldw, oldh);
    }

    @Override
    protected void onDraw(Canvas canvas) {

        if (bIsClear) {
            int width = DisplayUtil.DPFromPixel(mContext, 480);
            int height = DisplayUtil.DPFromPixel(mContext, 480);

            mBitmap.recycle();
            mBitmap = null;
            mBitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888);
            mCanvas.setBitmap(mBitmap);
            bIsClear = false;
        }

        canvas.drawBitmap(mBitmap, 0, 0, mBitmapPaint);
        if (bIsDrawStart) {
            if (mX != 0 && mY != 0) {
                canvas.drawPath(mPath, mPaint);
                canvas.drawPoint(mX, mY, mPaint);
            }
        }
    }

    public void clearText() {
        bIsClear = true;
        bIsDrawStart = false;
        bIsTouche = false;
        mPath.reset();
        mX = 0;
        mY = 0;

        invalidate();
    }

    private float mX, mY;
    private static final float TOUCH_TOLERANCE = 4;

    private void touch_start(float x, float y) {
        setIsTouche(false);
        mPath.reset();
        mPath.moveTo(x, y);
        mX = x;
        mY = y;

        bIsDrawStart = true;
    }

    private void touch_move(float x, float y) {
        float dx = Math.abs(x - mX);
        float dy = Math.abs(y - mY);
        if (dx >= TOUCH_TOLERANCE || dy >= TOUCH_TOLERANCE) {
            mPath.quadTo(mX, mY, (x + mX) / 2, (y + mY) / 2);
            mX = x;
            mY = y;
        }
    }

    private void touch_up() {
        // commit the path to our offscreen

        mPath.lineTo(mX, mY);
        mCanvas.drawPath(mPath, mPaint);
        mCanvas.drawPoint(mX, mY, mPaint);
        mPath.reset();
        bIsDrawStart = false;

        // kill this so we don't double draw
    }

    @Override
    public boolean onTouchEvent(MotionEvent event) {
        final int pointerCount = event.getPointerCount();
        setIsTouche(true);

        for (int p = 0; p < pointerCount; p++) {
            if (p == 0) {
                float x = event.getX();
                float y = event.getY();

                switch (event.getAction()) {
                    case MotionEvent.ACTION_DOWN:
                        this.getParent().requestDisallowInterceptTouchEvent(true);
                        touch_start(x, y);
                        invalidate();
                        break;
                    case MotionEvent.ACTION_MOVE:
                        if (mX != 0 && mY != 0) {
                            touch_move(x, y);
                            invalidate();
                        }
                        break;
                    case MotionEvent.ACTION_UP:
                        if (mX != 0 && mY != 0) {
                            this.getParent().requestDisallowInterceptTouchEvent(false);
                            touch_up();
                            invalidate();
                        }
                        break;
                }
            }
        }

        return true;
    }
}

File: app/src/main/java/com/giosis/util/qdrive/singapore/list/SigningView.kt
package com.giosis.util.qdrive.singapore.list

import android.content.Context
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.Path
import android.util.AttributeSet
import android.view.MotionEvent
import android.view.View
import com.giosis.util.qdrive.singapore.util.DisplayUtil

class SigningView : View {
    constructor(context: Context, attrs: AttributeSet?, defStyle: Int) : super(context, attrs, defStyle) {
        init()
    }

    companion object {
        private const val TOUCH_TOLERANCE = 4f
    }

    private var mBitmap: Bitmap? = null
    private var mCanvas: Canvas? = null
    private var mPath: Path? = null
    private var mBitmapPaint: Paint? = null
    private var mPaint: Paint? = null
    private var bIsClear = false
    private var bIsDrawStart = false
    var isTouch = false

    constructor(context: Context) : super(context) {
        init()
    }

    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs) {
        init()
    }

    private fun init() {

        mPaint = Paint()
        mPaint!!.isAntiAlias = true
        mPaint!!.isDither = true
        mPaint!!.color = -0x1000000
        mPaint!!.style = Paint.Style.STROKE
        mPaint!!.strokeJoin = Paint.Join.ROUND
        mPaint!!.strokeCap = Paint.Cap.ROUND
        mPaint!!.strokeWidth = 8f

        val width: Int = DisplayUtil.DPFromPixel(context, 480)
        val height: Int = DisplayUtil.DPFromPixel(context, 480)

        mBitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888)
        mCanvas = Canvas(mBitmap!!)
        mPath = Path()
        mBitmapPaint = Paint(Paint.DITHER_FLAG)
    }

    override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
        super.onSizeChanged(w, h, oldw, oldh)
    }

    override fun onDraw(canvas: Canvas) {
        if (bIsClear) {
            val width: Int = DisplayUtil.DPFromPixel(context, 480)
            val height: Int = DisplayUtil.DPFromPixel(context, 480)
            mBitmap!!.recycle()
            mBitmap = null
            mBitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888)
            mCanvas!!.setBitmap(mBitmap)
            bIsClear = false
        }

        canvas.drawBitmap(mBitmap!!, 0f, 0f, mBitmapPaint)
        if (bIsDrawStart) {
            if (mX != 0f && mY != 0f) {
                canvas.drawPath(mPath!!, mPaint!!)
                canvas.drawPoint(mX, mY, mPaint!!)
            }
        }
    }

    fun clearText() {
        bIsClear = true
        bIsDrawStart = false
        isTouch = false
        mPath!!.reset()
        mX = 0f
        mY = 0f
        invalidate()
    }

    private var mX = 0f
    private var mY = 0f
    private fun touch_start(x: Float, y: Float) {
        isTouch = false
        mPath!!.reset()
        mPath!!.moveTo(x, y)
        mX = x
        mY = y
        bIsDrawStart = true
    }

    private fun touch_move(x: Float, y: Float) {
        val dx = Math.abs(x - mX)
        val dy = Math.abs(y - mY)
        if (dx >= TOUCH_TOLERANCE || dy >= TOUCH_TOLERANCE) {
            mPath!!.quadTo(mX, mY, (x + mX) / 2, (y + mY) / 2)
            mX = x
            mY = y
        }
    }

    private fun touch_up() {
        // commit the path to our offscreen
        mPath!!.lineTo(mX, mY)
        mCanvas!!.drawPath(mPath!!, mPaint!!)
        mCanvas!!.drawPoint(mX, mY, mPaint!!)
        mPath!!.reset()
        bIsDrawStart = false
        // kill this so we don't double draw
    }

    override fun onTouchEvent(event: MotionEvent): Boolean {
        val pointerCount = event.pointerCount
        isTouch = true
        for (p in 0 until pointerCount) {
            if (p == 0) {
                val x = event.x
                val y = event.y
                when (event.action) {
                    MotionEvent.ACTION_DOWN -> {
                        this.parent.requestDisallowInterceptTouchEvent(true)
                        touch_start(x, y)
                        invalidate()
                    }
                    MotionEvent.ACTION_MOVE -> if (mX != 0f && mY != 0f) {
                        touch_move(x, y)
                        invalidate()
                    }
                    MotionEvent.ACTION_UP -> if (mX != 0f && mY != 0f) {
                        this.parent.requestDisallowInterceptTouchEvent(false)
                        touch_up()
                        invalidate()
                    }
                }
            }
        }
        return true
    }

}